File: EntativaAndroid/app/src/main/kotlin/com/entativa/network/AuthAPIClient.kt
package com.entativa.network

import android.content.Context
import android.content.SharedPreferences
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Authentication API Client for Entativa Backend
 * Enterprise-grade implementation with secure token storage
 */
class EntativaAuthAPIClient(private val context: Context) {
    
    companion object {
        private const val BASE_URL_DEBUG = "http://10.0.2.2:8001/api/v1"
        private const val BASE_URL_PRODUCTION = "https://api.entativa.com/api/v1"
        private const val TOKEN_KEY = "auth_token"
        
        @Volatile
        private var INSTANCE: EntativaAuthAPIClient? = null
        
        fun getInstance(context: Context): EntativaAuthAPIClient {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: EntativaAuthAPIClient(context.applicationContext).also {
                    INSTANCE = it
                }
            }
        }
    }
    
    private val baseUrl = if (BuildConfig.DEBUG) BASE_URL_DEBUG else BASE_URL_PRODUCTION
    
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()
    
    private val gson = Gson()
    
    private val securePrefs: SharedPreferences by lazy {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        
        EncryptedSharedPreferences.create(
            context,
            "entativa_secure_prefs",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }
    
    // MARK: - Data Models
    
    data class SignUpRequest(
        @SerializedName("first_name") val firstName: String,
        @SerializedName("last_name") val lastName: String,
        val email: String,
        val password: String,
        val birthday: String,
        val gender: String
    )
    
    data class LoginRequest(
        @SerializedName("email_or_username") val emailOrUsername: String,
        val password: String
    )
    
    data class AuthResponse(
        val success: Boolean,
        val message: String,
        val data: AuthData?
    )
    
    data class AuthData(
        val user: User,
        @SerializedName("access_token") val accessToken: String,
        @SerializedName("token_type") val tokenType: String,
        @SerializedName("expires_in") val expiresIn: Int
    )
    
    data class User(
        val id: String,
        @SerializedName("first_name") val firstName: String,
        @SerializedName("last_name") val lastName: String,
        val email: String,
        val username: String,
        val birthday: String?,
        val gender: String?,
        @SerializedName("profile_picture_url") val profilePictureUrl: String?,
        @SerializedName("cover_photo_url") val coverPhotoUrl: String?,
        @SerializedName("is_active") val isActive: Boolean,
        @SerializedName("created_at") val createdAt: String
    )
    
    data class ErrorResponse(
        val success: Boolean,
        val error: String,
        val details: Map<String, List<String>>?
    )
    
    // MARK: - API Methods
    
    /**
     * Sign up a new user
     */
    suspend fun signUp(
        firstName: String,
        lastName: String,
        email: String,
        password: String,
        birthday: String,
        gender: String
    ): Result<AuthResponse> = withContext(Dispatchers.IO) {
        try {
            val requestBody = SignUpRequest(
                firstName = firstName,
                lastName = lastName,
                email = email,
                password = password,
                birthday = birthday,
                gender = gender
            )
            
            val json = gson.toJson(requestBody)
            val body = json.toRequestBody("application/json".toMediaType())
            
            val request = Request.Builder()
                .url("$baseUrl/auth/signup")
                .post(body)
                .build()
            
            val response = client.newCall(request).execute()
            
            if (response.isSuccessful) {
                val authResponse = gson.fromJson(response.body?.string(), AuthResponse::class.java)
                authResponse.data?.accessToken?.let { saveToken(it) }
                Result.success(authResponse)
            } else {
                val errorResponse = gson.fromJson(response.body?.string(), ErrorResponse::class.java)
                Result.failure(AuthException(errorResponse.error))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Log in an existing user
     */
    suspend fun login(
        emailOrUsername: String,
        password: String
    ): Result<AuthResponse> = withContext(Dispatchers.IO) {
        try {
            val requestBody = LoginRequest(
                emailOrUsername = emailOrUsername,
                password = password
            )
            
            val json = gson.toJson(requestBody)
            val body = json.toRequestBody("application/json".toMediaType())
            
            val request = Request.Builder()
                .url("$baseUrl/auth/login")
                .post(body)
                .build()
            
            val response = client.newCall(request).execute()
            
            if (response.isSuccessful) {
                val authResponse = gson.fromJson(response.body?.string(), AuthResponse::class.java)
                authResponse.data?.accessToken?.let { saveToken(it) }
                Result.success(authResponse)
            } else {
                val errorResponse = gson.fromJson(response.body?.string(), ErrorResponse::class.java)
                Result.failure(AuthException(errorResponse.error))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Get current authenticated user
     */
    suspend fun getCurrentUser(): Result<User> = withContext(Dispatchers.IO) {
        try {
            val token = getToken() ?: return@withContext Result.failure(AuthException("Not authenticated"))
            
            val request = Request.Builder()
                .url("$baseUrl/auth/me")
                .header("Authorization", "Bearer $token")
                .get()
                .build()
            
            val response = client.newCall(request).execute()
            
            if (response.isSuccessful) {
                val userResponse = gson.fromJson(response.body?.string(), UserResponse::class.java)
                Result.success(userResponse.data)
            } else {
                Result.failure(AuthException("Failed to get user"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Log out current user
     */
    suspend fun logout(): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val token = getToken()
            
            if (token != null) {
                val request = Request.Builder()
                    .url("$baseUrl/auth/logout")
                    .header("Authorization", "Bearer $token")
                    .post("".toRequestBody())
                    .build()
                
                client.newCall(request).execute()
            }
            
            deleteToken()
            Result.success(Unit)
        } catch (e: Exception) {
            deleteToken() // Clear token even if request fails
            Result.success(Unit)
        }
    }
    
    // MARK: - Token Management
    
    fun saveToken(token: String) {
        securePrefs.edit().putString(TOKEN_KEY, token).apply()
    }
    
    fun getToken(): String? {
        return securePrefs.getString(TOKEN_KEY, null)
    }
    
    fun deleteToken() {
        securePrefs.edit().remove(TOKEN_KEY).apply()
    }
    
    fun hasToken(): Boolean {
        return getToken() != null
    }
    
    // Helper class for user response
    data class UserResponse(
        val success: Boolean,
        val data: User
    )
}

/**
 * Custom exception for authentication errors
 */
class AuthException(message: String) : Exception(message)

// BuildConfig placeholder - should be generated by build system
object BuildConfig {
    const val DEBUG = true
}

File: EntativaiOS/Services/API/AuthAPIClient.swift
import Foundation
import Combine

/// Authentication API client for Entativa backend
class AuthAPIClient {
    static let shared = AuthAPIClient()
    
    private let baseURL: String
    private let session: URLSession
    
    private init() {
        #if DEBUG
        self.baseURL = "http://localhost:8001/api/v1"
        #else
        self.baseURL = "https://api.entativa.com/api/v1"
        #endif
        
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 30
        configuration.timeoutIntervalForResource = 60
        self.session = URLSession(configuration: configuration)
    }
    
    // MARK: - Models
    
    struct SignUpRequest: Codable {
        let firstName: String
        let lastName: String
        let email: String
        let password: String
        let birthday: String
        let gender: String
        
        enum CodingKeys: String, CodingKey {
            case firstName = "first_name"
            case lastName = "last_name"
            case email
            case password
            case birthday
            case gender
        }
    }
    
    struct LoginRequest: Codable {
        let emailOrUsername: String
        let password: String
        
        enum CodingKeys: String, CodingKey {
            case emailOrUsername = "email_or_username"
            case password
        }
    }
    
    struct AuthResponse: Codable {
        let success: Bool
        let message: String
        let data: AuthData?
        
        struct AuthData: Codable {
            let user: User
            let accessToken: String
            let tokenType: String
            let expiresIn: Int
            
            enum CodingKeys: String, CodingKey {
                case user
                case accessToken = "access_token"
                case tokenType = "token_type"
                case expiresIn = "expires_in"
            }
        }
        
        struct User: Codable {
            let id: String
            let firstName: String
            let lastName: String
            let email: String
            let username: String
            let birthday: String?
            let gender: String?
            let profilePictureUrl: String?
            let coverPhotoUrl: String?
            let isActive: Bool
            let createdAt: String
            
            enum CodingKeys: String, CodingKey {
                case id
                case firstName = "first_name"
                case lastName = "last_name"
                case email
                case username
                case birthday
                case gender
                case profilePictureUrl = "profile_picture_url"
                case coverPhotoUrl = "cover_photo_url"
                case isActive = "is_active"
                case createdAt = "created_at"
            }
        }
    }
    
    struct ErrorResponse: Codable {
        let success: Bool
        let error: String
        let details: [String: [String]]?
    }
    
    // MARK: - API Methods
    
    /// Sign up a new user
    func signUp(
        firstName: String,
        lastName: String,
        email: String,
        password: String,
        birthday: Date,
        gender: String
    ) async throws -> AuthResponse {
        let endpoint = "\(baseURL)/auth/signup"
        
        guard let url = URL(string: endpoint) else {
            throw AuthError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        
        let signUpRequest = SignUpRequest(
            firstName: firstName,
            lastName: lastName,
            email: email,
            password: password,
            birthday: dateFormatter.string(from: birthday),
            gender: gender
        )
        
        request.httpBody = try JSONEncoder().encode(signUpRequest)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw AuthError.invalidResponse
        }
        
        if httpResponse.statusCode == 200 || httpResponse.statusCode == 201 {
            let authResponse = try JSONDecoder().decode(AuthResponse.self, from: data)
            
            // Store token
            if let token = authResponse.data?.accessToken {
                try KeychainManager.shared.save(token: token)
            }
            
            return authResponse
        } else {
            let errorResponse = try? JSONDecoder().decode(ErrorResponse.self, from: data)
            throw AuthError.serverError(errorResponse?.error ?? "Sign up failed")
        }
    }
    
    /// Log in an existing user
    func login(emailOrUsername: String, password: String) async throws -> AuthResponse {
        let endpoint = "\(baseURL)/auth/login"
        
        guard let url = URL(string: endpoint) else {
            throw AuthError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let loginRequest = LoginRequest(
            emailOrUsername: emailOrUsername,
            password: password
        )
        
        request.httpBody = try JSONEncoder().encode(loginRequest)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw AuthError.invalidResponse
        }
        
        if httpResponse.statusCode == 200 {
            let authResponse = try JSONDecoder().decode(AuthResponse.self, from: data)
            
            // Store token
            if let token = authResponse.data?.accessToken {
                try KeychainManager.shared.save(token: token)
            }
            
            return authResponse
        } else {
            let errorResponse = try? JSONDecoder().decode(ErrorResponse.self, from: data)
            throw AuthError.serverError(errorResponse?.error ?? "Login failed")
        }
    }
    
    /// Get current authenticated user
    func getCurrentUser() async throws -> AuthResponse.User {
        let endpoint = "\(baseURL)/auth/me"
        
        guard let url = URL(string: endpoint) else {
            throw AuthError.invalidURL
        }
        
        guard let token = try? KeychainManager.shared.getToken() else {
            throw AuthError.notAuthenticated
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw AuthError.invalidResponse
        }
        
        if httpResponse.statusCode == 200 {
            struct UserResponse: Codable {
                let success: Bool
                let data: AuthResponse.User
            }
            let userResponse = try JSONDecoder().decode(UserResponse.self, from: data)
            return userResponse.data
        } else {
            throw AuthError.notAuthenticated
        }
    }
    
    /// Log out current user
    func logout() async throws {
        let endpoint = "\(baseURL)/auth/logout"
        
        guard let url = URL(string: endpoint) else {
            throw AuthError.invalidURL
        }
        
        guard let token = try? KeychainManager.shared.getToken() else {
            // Already logged out
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        
        let (_, response) = try await session.data(for: request)
        
        // Clear token regardless of server response
        try KeychainManager.shared.deleteToken()
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw AuthError.invalidResponse
        }
        
        if httpResponse.statusCode != 200 {
            print("Warning: Logout returned status \(httpResponse.statusCode)")
        }
    }
}

// MARK: - Auth Errors

enum AuthError: LocalizedError {
    case invalidURL
    case invalidResponse
    case notAuthenticated
    case serverError(String)
    case networkError(Error)
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid API URL"
        case .invalidResponse:
            return "Invalid server response"
        case .notAuthenticated:
            return "Not authenticated. Please log in."
        case .serverError(let message):
            return message
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        }
    }
}

// MARK: - Keychain Manager

class KeychainManager {
    static let shared = KeychainManager()
    private let service = "com.entativa.app"
    private let account = "authToken"
    
    private init() {}
    
    func save(token: String) throws {
        let data = token.data(using: .utf8)!
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecValueData as String: data
        ]
        
        // Delete any existing item
        SecItemDelete(query as CFDictionary)
        
        // Add new item
        let status = SecItemAdd(query as CFDictionary, nil)
        
        guard status == errSecSuccess else {
            throw KeychainError.saveFailed
        }
    }
    
    func getToken() throws -> String {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        guard status == errSecSuccess,
              let data = result as? Data,
              let token = String(data: data, encoding: .utf8) else {
            throw KeychainError.notFound
        }
        
        return token
    }
    
    func deleteToken() throws {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account
        ]
        
        let status = SecItemDelete(query as CFDictionary)
        
        guard status == errSecSuccess || status == errSecItemNotFound else {
            throw KeychainError.deleteFailed
        }
    }
}

enum KeychainError: Error {
    case saveFailed
    case notFound
    case deleteFailed
}
